### PR TITLE
Use same retrigger logic for Multiplexer and InputShifter

### LINKED ISSUE #200: Mega get's sometimes not properly identified
Sometimes the Mega (and I expect the same for the Uno) get's not properly identified from the connector. That means that:
* eeprom config
* serial number
* devices
can be misread (not always). This was found from @Jaime Leon and is reported on Discord.
He found also, that all switches put into OFF position helped for properly booting.

It seems that this behaviour is related to the initilisation of the buttons. Every button is initilised woith `1`, so after startingup for each ON button a status change is detected and a serial message is generated. For multiple buttons multiple messages are generated. As this comes directly after the GUI has connected (reset via DTR for Mega/Uno) and the connector asks for the board info's, it seems that this disturbs the correct identification of the aboved mentioned infos.
A first test from Jaime Leon of initialisation the buttons with the actual status solved this problem. As no status change is identified on startup no serial message gets generated.

The same applies to:
* input shifter
* multiplexer
* analog in
 For analog in this is fixed by PR #199 where the average buffer is filled on startup. For the input shifter and buttons I have already prepared the changes.
For the multiplexer I would like to ask @GioCC to prepare the changes as I am unsure what the best way would be.

### PR DESCRIPTION
**Pull request by GioCC**
Changes:

- [x] Added value initialization for digital input multiplexers
- [x] Introduce enum DO_TRIGGER and DONT_TRIGGER for both device types
- [x] Introduce poll method for both device types to reduce duplicate code
- [x] use same trigger method logic 

## Description of changes
relates to #201 
fixes #200

### CODE REVIEW

File: src/MF_InputShifter/MFInputShifter.cpp [hunk 1]
```diff
@@ -10,146 +10,136 @@ inputShifterEvent MFInputShifter::_inputHandler = NULL;
...
 }
 
 void MFInputShifter::detach()
 {
-  if (!_initialized)
-    return;
-  _initialized = false;
+    if (!_initialized)
+        return;
+    _initialized = false;
```
Comment (elral) [inline src/MF_InputShifter/MFInputShifter.cpp:142]: Isn't just `_initialized=false;` enough?

File: src/MF_InputShifter/MFInputShifter.cpp [hunk 2]
```diff
@@ -10,146 +10,136 @@ inputShifterEvent MFInputShifter::_inputHandler = NULL;
...
 {
     if (state == LOW && _inputHandler != NULL) {
-    (*_inputHandler)(inputShifterOnPress, pin, _name);
+        (*_inputHandler)(inputShifterOnPress, pin, _name);
     } else if (_inputHandler != NULL) {
-    (*_inputHandler)(inputShifterOnRelease, pin, _name);
-  }
+        (*_inputHandler)(inputShifterOnRelease, pin, _name);
+    }
 }
```
Comment (elral) [inline src/MF_InputShifter/MFInputShifter.cpp:131]: Would it make sense to use the same code as for the multiplexer?
`if (!_inputHandler) return;
(*_inputHandler)((state ? inputShifterOnRelease : inputShifterOnPress), pin, _name);`

**Comment by elral** [comment]
Additionally we have still the topic with auto formatting the code.
I still don't know if something has to be add to my personal platformIO or if a general setting has to be released.

File: src/MF_InputShifter/MFInputShifter.cpp (hunk 2 above)
Comment (GioCC) [reply src/MF_InputShifter/MFInputShifter.cpp:131]: Yes it would. Didn't go over this section of code.

**Comment by GioCC** [comment]
> 



> Additionally we have still the topic with auto formatting the code. I still don't know if something has to be add to my personal platformIO or if a general setting has to be released.

I believe I'm using the last version that was basically agreed upon; I don't recall if it was ever been made official by incorporating it in the project (but I guess we both would have found it in the git checkout if this was the case).

**Comment by DocMoebiuz** [comment]
> > 
> 
> > Additionally we have still the topic with auto formatting the code. I still don't know if something has to be add to my personal platformIO or if a general setting has to be released.
> 
> I believe I'm using the last version that was basically agreed upon; I don't recall if it was ever been made official by incorporating it in the project (but I guess we both would have found it in the git checkout if this was the case).

you have to check indentation settings, the code used to be 2 spaces per indent, your settings seems to be 4.

**Comment by GioCC** [comment]
> > > 
> > 
> > 
> > > Additionally we have still the topic with auto formatting the code. I still don't know if something has to be add to my personal platformIO or if a general setting has to be released.
> > 
> > 
> > I believe I'm using the last version that was basically agreed upon; I don't recall if it was ever been made official by incorporating it in the project (but I guess we both would have found it in the git checkout if this was the case).
> 
> you have to check indentation settings, the code used to be 2 spaces per indent, your settings seems to be 4. Maybe I am wrong, I looked at the code... majority is now 4 spaces.
> 
> I looked back in history, it used to be 2 spaces only. So yes, we have to have a setting that we are all using. Which file is it? Maybe it is not committed to the repository?
> 
> "editor.formatOnSave": true - this is also the reason that we see the format changing.

https://github.com/MobiFlight/MobiFlight-FirmwareSource/blob/a57611caa55660e2459c5825715226c82dab1598/.vscode/settings.json#L2

**Comment by elral** [comment]
> It's there, so I assume it was agreed upon at some point. Anyway it's what affects my saves.

Ah, yes, I now remember. It was agreed on this, but I also remember that this does not "work" for autosave which I have enabled (and forgot completly). I have to actively save the file. I will do my best to do it before commiting new files...

**Comment by elral** [comment]
I just doubled checked the formatting, and indeed the file will not be formatted if it is not actively saved (CTRL-S or from menu).
As an example see the attached screen shot. Only with pressing CTRL-S the file is formatted what I havn't done before because I have Autosave enabled. Sorry for this.
But I am still wondering why in this file the TabWidth is still 2 and not 4 according the settings.

image

**Comment by GioCC** [comment]
> I just doubled checked the formatting, and indeed the file will not be formatted if it is not actively saved (CTRL-S or from menu). As an example see the attached screen shot. Only with pressing CTRL-S the file is formatted what I havn't done before because I have Autosave enabled. Sorry for this. 

Thinking about it, it totally makes sense, because otherwise autosave would possibly reformat your code "under your feet" while you are working... (I personally refrain from using autosave, so I never noticed)

> But I am still wondering why in this file the TabWidth is still 2 and not 4 according the settings.

This file had probably not been modified, thus saved, in a long time.

I think it would make sense to make a specific commit only with reformatting applied to all files, that would avoid recurring problems.
 
> /edit: sometimes it is good to check all open issues -> #159 ;) We should continue the discussion of formatting in this issue...

Yes. Often small (meaning, quick to solve) issues linger around for long.
Though the presence of the config file seems to suggest that the issue was decided at some point in time?

**Comment by neilenns** [comment]
This is the solution to making format happen on auto save: https://marketplace.visualstudio.com/items?itemName=pucelle.run-on-save

Vs code doesn't support it natively despite many requests.


________________________________

**Comment by neilenns** [comment]
Here's a sample settings.json snippet for that extension:

```
"runOnSave.commands": [
    {
        "match": ".ps1",
        "command": "editor.action.formatDocument",
        "runIn": "vscode"
    }
]
```

**Comment by neilenns** [comment]
Or try this: https://marketplace.visualstudio.com/items?itemName=BdSoftware.format-on-auto-save

**Comment by DocMoebiuz** [comment]
> Or try this: https://marketplace.visualstudio.com/items?itemName=BdSoftware.format-on-auto-save

for me i have that setting in the cpp extension which is auto format on save but not auto format on auto save.

**Comment by elral** [comment]
@neilenns thanks for the informations! I will give them a try.

**Comment by elral** [comment]
> (Though the presence of the config file seems to suggest that the matter was decided at some point in time?)

At least it was merged.

**Comment by DocMoebiuz** [comment]
@Gioc

> Fix: Added initial value setting also for digital input multiplexers Mod: Refactored duplicate code in input shifter handling
> 
> ## Description of changes
> Completed fixes for #200 as follow-up to PR #201

@GioCC please add some details about this change, so that it is easier for me to understand the motiviation, the improvement and the scope. What is the follow up on PR #201 exactly?

**Review by DocMoebiuz — CHANGES_REQUESTED** [review]
Hi @GioCC thanks for helping with this. For an easier review, please consider the following points:

* Provide a clear and simple PR title that summarizes what the PR does, e.g. in this case
  * **Use same retrigger logic for Multiplexer and InputShifter**

* Provide a short outline of the significant changes, e.g. as a checkbox list
  - [ ] Introduce enum DO_TRIGGER and DONT_TRIGGER for both device types
  - [ ] Introduce poll method for both device types to reduce duplicate code
  - [ ] use same trigger method logic 

That would help me a lot to faster grasp the intention of the PR and reviewing the code.

**Review by DocMoebiuz — APPROVED** [review]
I made the changes for you in the comment.
Thanks @GioCC

**Pull request merged**